File: fushi/test/pages/popup_pitch_dedup_patterns_ipa_test.js
// BUG-2397 behavior test: 「音调去重」对 pattern 式音调与 IPA 完全不生效。
//
// 用户报告：设置里「Deduplicate pitch accents」开着，音高区照样出现一模一样的重复行。
//
// 根因：createPitchSection 的去重分支只把**数字位置**（pitchPositions）收进 seen，
// `patterns`（"heiban" 等 pattern 式音调）与 `transcriptions`（IPA 音标）一概不参与去重，
// 而且「保活守卫」用的是「原始字段非空」——于是第二本词典即使一个字都不新，也照样整行
// 渲染出来。上游的 mergeIdenticalPitchGroups 只能接住「整份 payload 全等」的那一种；
// 两本词典只要在任意一个字段上差一点（一本带 IPA、一本不带），合并就不成立，重复
// 全部落到去重这一步，而这一步以前对这两类视而不见。
//
// 本测试 EXECUTES 真实的 popup.js（vm + 极简假 DOM），驱动真实的 createPitchSection，
// 然后在产出的元素树里数「同一个可见条目出现了几次」。把三个 seen 里的任何一个退回
// 原来的单一 seen，对应 case 立刻变红。
//
// Cases covered:
//   1. 两本词典同标 pattern `heiban`、其中一本另带 IPA（合并不成立）→ `[heiban]` 只出现一次。
//   2. 两本词典给出同一段 IPA、数字位置不同（合并不成立）→ 该 IPA 只出现一次，两个位置都在。
//   3. 去重关闭时，重复的 pattern / IPA 全部保留（开关必须仍然有意义，非恒真）。
//   4. 不回归 TODO-688：唯一的 IPA-only 词典（无数字位置）仍然渲染出它的 IPA。
//   5. 不回归 BUG-2122：五本同标 [1] 的词典塌成 1 行、5 枚来源药丸。
//
// Run: node fushi/test/pages/popup_pitch_dedup_patterns_ipa_test.js
// (also driven from popup_pitch_dedup_patterns_ipa_test.dart so it executes inside
//  `flutter test`).

const assert = require('assert');
const {
  loadPopup,
  collectByClass,
  collectText,
} = require('./_popup_dom_host.js');

function dict(name, positions, patterns, transcriptions) {
  return {
    dictionary: name,
    pitchPositions: positions || [],
    patterns: patterns || [],
    transcriptions: transcriptions || [],
  };
}

function render(pitches, dedup, reading) {
  const sb = loadPopup();
  sb.window.deduplicatePitchAccents = dedup;
  return sb.window.__test.createPitchSection(pitches, reading || 'ギター');
}

// 整棵树里某段可见文本出现了几次。pattern 式音调渲染成裸文本节点 `[heiban]`，
// IPA 渲染成 .pitch-transcription-tag 的 `[ipa]`——数「用户眼睛看到几次」，
// 所以统一在文本层数，而不是数元素类型。
function occurrences(section, needle) {
  const text = collectText(section);
  let count = 0;
  let from = 0;
  for (;;) {
    const at = text.indexOf(needle, from);
    if (at < 0) return count;
    count += 1;
    from = at + needle.length;
  }
}

function ipaTags(section) {
  return collectByClass(section, 'pitch-transcription-tag').map(n => n.textContent);
}

(function run() {
  // Case 1: 用户报告的主形态——两本词典都说 heiban，其中一本另带 IPA。
  // 整份 payload 不全等 ⇒ mergeIdenticalPitchGroups 合并不了 ⇒ 全靠去重这一步。
  {
    const section = render(
      [dict('d1', [], ['heiban']), dict('d2', [], ['heiban'], ['ɡiꜜtaː'])],
      true,
    );
    assert.ok(section, 'pattern-only pitch groups must still render a section');
    const heiban = occurrences(section, '[heiban]');
    assert.strictEqual(heiban, 1,
      'with dedup ON, the same pattern accent must be shown ONCE across '
        + 'dictionaries; got ' + heiban + ' copies');
    // 第二本词典独有的 IPA 不能被误杀——去重删的是重复，不是整行。
    assert.deepStrictEqual(ipaTags(section), ['[ɡiꜜtaː]'],
      'the second dictionary\'s unique IPA must survive dedup');
  }

  // Case 2: 两本 IPA 词典给出同一串音标，数字位置不同（同样合并不了）。
  {
    const section = render(
      [dict('d1', [1], [], ['ɡiꜜtaː']), dict('d2', [0], [], ['ɡiꜜtaː'])],
      true,
    );
    assert.deepStrictEqual(ipaTags(section), ['[ɡiꜜtaː]'],
      'with dedup ON, an identical IPA transcription must be shown ONCE');
    // 两个数字位置都是新的，必须都留下（去重不是「只留第一本词典」）。
    const text = collectText(section);
    assert.ok(text.includes('[1]') && text.includes('[0]'),
      'both distinct pitch positions must survive; got: ' + text);
  }

  // Case 3: 非恒真——关掉开关，重复必须回来。否则上面两条即使去重代码整块删掉也能过。
  {
    const section = render(
      [dict('d1', [], ['heiban']), dict('d2', [], ['heiban'], ['ɡiꜜtaː'])],
      false,
    );
    assert.strictEqual(occurrences(section, '[heiban]'), 2,
      'with dedup OFF the duplicate pattern must still be rendered twice '
        + '(otherwise the setting would be meaningless)');
  }

  // Case 4: TODO-688 不回归——纯 IPA 词典（无数字位置）在去重打开时仍要渲染 IPA。
  {
    const section = render([dict('d1', [1]), dict('ipa', [], [], ['neꜜko'])], true);
    assert.deepStrictEqual(ipaTags(section), ['[neꜜko]'],
      'an IPA-only dictionary must still render under dedup (TODO-688)');
  }

  // Case 5: BUG-2122 不回归——五本同标 [1] 塌成一行、五枚来源药丸。
  {
    const five = ['d1', 'd2', 'd3', 'd4', 'd5'].map(n => dict(n, [1]));
    const section = render(five, true);
    assert.strictEqual(collectByClass(section, 'pitch-group').length, 1,
      'five dictionaries agreeing on [1] must stay collapsed into one row');
    assert.strictEqual(collectByClass(section, 'pitch-dict-label').length, 5,
      'all five source labels must survive (BUG-2122)');
  }

  console.log('all assertions passed');
})();

File: fushi/test/pages/popup_pitch_dedup_patterns_ipa_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

/// BUG-2397：「音调去重」对 pattern 式音调与 IPA 完全不生效，且在浏览器扩展里
/// **从来没生效过**。
///
/// 用户报告：设置里 Deduplicate pitch accents 开着，音高区照样出现一模一样的重复行。
///
/// 两条独立根因：
/// ① `createPitchSection` 的去重分支只把**数字位置**收进 seen，`patterns`
///    （"heiban" 等 pattern 式音调）与 `transcriptions`（IPA）一概不参与；保活守卫
///    又用「原始字段非空」，于是第二本词典一个字都不新也照样整行渲染。上游的
///    `mergeIdenticalPitchGroups` 只接得住「整份 payload 全等」那一种，两本词典
///    只要差一个字段（一本带 IPA、一本不带）就合并不了，重复全落到去重这一步。
/// ② 浏览器扩展里 `window.deduplicatePitchAccents` **从未有人赋值**：这个偏好没被
///    放进 app 下发给扩展的 theme 通道，扩展 content.js 也没有消费它，于是恒
///    undefined = falsy，去重分支恒不执行——用户在 app 里怎么设都没用。
///
/// 三层守护：
/// ① 行为级——用 Node 真执行 popup.js 的 `createPitchSection`（见同名 .js）。
/// ② 源码级——扫描**三份 popup.js 镜像**，钉住三类可见条目各自入 seen、保活判据
///    用去重后的结果。无 node 时也能守住回归。
/// ③ 接线级——钉住 app 侧下发 `--fushi-dedup-pitch` 与**两份** content.js 消费它，
///    否则根因②会悄悄回来（那条链路上没有任何行为测试能覆盖）。
void main() {
  const Map<String, String> jsMirrors = <String, String>{
    'app popup': 'assets/popup/popup.js',
    'extension vendor (assets)': 'assets/browser_extension/vendor/popup.js',
    'extension vendor (tools)': '../tools/browser-extension/vendor/popup.js',
  };
  const Map<String, String> contentMirrors = <String, String>{
    'extension content (assets)': 'assets/browser_extension/content.js',
    'extension content (tools)': '../tools/browser-extension/content.js',
  };

  test(
    'pattern accents and IPA transcriptions are deduplicated across dictionaries '
    '(executes createPitchSection via node)',
    () async {
      final String? nodeExe = _resolveNode();
      if (nodeExe == null) {
        markTestSkipped(
            'node not found on PATH; skipping JS behavior execution');
        return;
      }

      final File jsTest =
          File('test/pages/popup_pitch_dedup_patterns_ipa_test.js');
      expect(
        jsTest.existsSync(),
        isTrue,
        reason: 'behavior harness ${jsTest.path} must exist',
      );

      final ProcessResult result = await Process.run(
        nodeExe,
        <String>[jsTest.path],
        workingDirectory: Directory.current.path,
      );

      expect(
        result.exitCode,
        0,
        reason: 'pitch dedup (patterns / IPA) JS behavior test failed.\n'
            'stdout:\n${result.stdout}\nstderr:\n${result.stderr}',
      );
      expect(
        result.stdout.toString(),
        contains('all assertions passed'),
        reason: 'behavior harness must reach its success marker',
      );
    },
  );

  jsMirrors.forEach((String name, String relPath) {
    test('[$name] pitch dedup covers positions, patterns AND transcriptions',
        () {
      final File file = File(relPath);
      expect(file.existsSync(), isTrue, reason: '$relPath must exist');
      final String js = file.readAsStringSync();

      final int dedup = js.indexOf('if (window.deduplicatePitchAccents)');
      expect(dedup, greaterThanOrEqualTo(0),
          reason: 'the dedup branch must exist');

      // 三类可见条目各有自己的 seen。少一个 = 那一类永远不去重（本 bug 的原状）。
      for (final String seen in <String>[
        'const seenPositions = new Set();',
        'const seenPatterns = new Set();',
        'const seenTranscriptions = new Set();',
      ]) {
        expect(
          js.indexOf(seen, dedup),
          greaterThan(dedup),
          reason: 'the dedup branch must keep a dedicated set: $seen',
        );
      }

      // 非恒真：三个 Set 建了却没人过滤等于没改。三类都必须真的按 seen 过滤。
      for (final String filter in <String>[
        '.filter(pos => !seenPositions.has(pos))',
        '.filter(p => !seenPatterns.has(p))',
        '.filter(ipa => !seenTranscriptions.has(ipa))',
      ]) {
        expect(
          js.indexOf(filter, dedup),
          greaterThan(dedup),
          reason: 'each class of visible entry must be filtered: $filter',
        );
      }

      // 保活判据必须是「去重后还剩东西」。写成原始字段非空（`group.patterns?.length`）
      // 的那一刻，一整行已经显示过的 pattern / IPA 又会被画第二遍。
      final int guard = js.indexOf(
        'if (unique.length > 0 || uniquePatterns.length > 0 || '
        'uniqueTranscriptions.length > 0)',
        dedup,
      );
      expect(
        guard,
        greaterThan(dedup),
        reason: 'the keep-alive guard must test the DEDUPED entries, not the '
            'raw fields — testing the raw fields is exactly BUG-2397',
      );
    });
  });

  test('the pitch dedup preference is wired through to the browser extension',
      () {
    // 根因②的守卫。app 侧把偏好放进查词响应的 theme 字典（非 CSS 变量、仅 JS 消费，
    // 与 --fushi-instant-scroll 同法），扩展 content.js 据此设 popup.js 读的全局。
    final String appModel =
        File('lib/src/models/app_model.dart').readAsStringSync();
    expect(
      appModel,
      contains("'--fushi-dedup-pitch': deduplicatePitchAccents ? '1' : '0',"),
      reason:
          'app_model must send the pitch-dedup preference to the extension; '
          'without it window.deduplicatePitchAccents is undefined in the '
          'browser and the dedup branch never runs',
    );

    contentMirrors.forEach((String name, String relPath) {
      final File file = File(relPath);
      expect(file.existsSync(), isTrue, reason: '$relPath must exist');
      expect(
        file.readAsStringSync(),
        contains(
            "window.deduplicatePitchAccents = theme['--fushi-dedup-pitch'] === '1';"),
        reason: '[$name] content.js must publish the preference on the shared '
            'window that popup.js reads',
      );
    });
  });
}

/// Resolve a usable `node` executable, returning null when none is on PATH.
String? _resolveNode() {
  final List<String> candidates =
      Platform.isWindows ? <String>['node.exe', 'node'] : <String>['node'];
  for (final String name in candidates) {
    try {
      final ProcessResult probe = Process.runSync(name, <String>['--version']);
      if (probe.exitCode == 0) {
        return name;
      }
    } on ProcessException {
      // Not found; try next candidate.
    }
  }
  return null;
}
